test/Microsoft.ML.TimeSeries.Tests/TimeSeriesDirectApi.cs

```diff
@@ -515,5 +519,155 @@ public void AnomalyDetectionWithSrCnn()
                 k += 1;
             }
         }
+
+        private class RootCauseLocalizationData
+        {
+            [RootCauseLocalizationInputType]
+            public RootCauseLocalizationInput Input { get; set; }
+
+            public RootCauseLocalizationData()
+            {
+                Input = null;
+            }
+
+            public RootCauseLocalizationData(DateTime anomalyTimestamp, Dictionary<string, string> anomalyDimensions, List<MetricSlice> slices, AggregateType aggregateteType, string aggregateSymbol)
+            {
+                Input = new RootCauseLocalizationInput(anomalyTimestamp, anomalyDimensions, slices, aggregateteType, aggregateSymbol);
+            }
+        }
+
+        private class RootCauseLocalizationTransformedData
+        {
+            [RootCauseType()]
+            public RootCause RootCause { get; set; }
+
+            public RootCauseLocalizationTransformedData()
+            {
+                RootCause = null;
+            }
+        }
+
+        [Fact]
+        public void RootCauseLocalizationWithDT()
```

Comment:
>RootCauseLocalizationWithDT [](start = 20, length = 27)

Please also test serialization/deserialization.

Reply:
Could you help explain more, I did't exactly understand the suggestion. Which part needs serialization/deserialization?

Reply:
> > RootCauseLocalizationWithDT [](start = 20, length = 27)
> 
> Please also test serialization/deserialization.

Could you help explain more, I did't exactly understand the suggestion. Which part needs serialization/deserialization?

Reply:
Part of the `DTRootCauseLocalizationTransformer` class is methods that save the model to a file, and that load the model back from a file. This test should exercise these methods, by using the `ml.Model.Save` and `ml.Model.Load` APIs, and checking that the values returned by the deserialized model are the same as the ones returned by the original model.

---
In reply to: [399045379](https://github.com/dotnet/machinelearning/pull/4925#discussion_r399045379) [](ancestors = 399045379)

Reply:
> Part of the `DTRootCauseLocalizationTransformer` class is methods that save the model to a file, and that load the model back from a file. This test should exercise these methods, by using the `ml.Model.Save` and `ml.Model.Load` APIs, and checking that the values returned by the deserialized model are the same as the ones returned by the original model.
> 
> In reply to: [399045379](https://github.com/dotnet/machinelearning/pull/4925#discussion_r399045379) [](ancestors = 399045379)

Do I need to save the model, I find in the document, the save menthod is used for training related data

Reply:
Sorry, I was unclear. You should save the **model**, not the data. Using this API:
[https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.Data/Model/ModelOperationsCatalog.cs#L112](https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.Data/Model/ModelOperationsCatalog.cs#L112)
You then need to load the model using this API:
[https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.Data/Model/ModelOperationsCatalog.cs#L211](https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.Data/Model/ModelOperationsCatalog.cs#L211)
and then run the loaded model on the data, and verify that it gives the same outputs as the model before serialization.

Here is a test that contains an example of how to save and load the model:
[https://github.com/dotnet/machinelearning/blob/master/test/Microsoft.ML.Tests/Transformers/NormalizerTests.cs#L961](https://github.com/dotnet/machinelearning/blob/master/test/Microsoft.ML.Tests/Transformers/NormalizerTests.cs#L961)

(the saving and loading is done in lines 969 and 970).

---
In reply to: [399795392](https://github.com/dotnet/machinelearning/pull/4925#discussion_r399795392) [](ancestors = 399795392)

Reply:
> Part of the `DTRootCauseLocalizationTransformer` class is methods that save the model to a file, and that load the model back from a file. This test should exercise these methods, by using the `ml.Model.Save` and `ml.Model.Load` APIs, and checking that the values returned by the deserialized model are the same as the ones returned by the original model.
> 
> In reply to: [399045379](https://github.com/dotnet/machinelearning/pull/4925#discussion_r399045379) [](ancestors = 399045379)

I have tried to use this logic to save and load model, 

            var modelPath = "temp.zip";
            //Save model to a file
            ml.Model.Save(model, data.Schema, modelPath);

            //Load model from a file
            ITransformer serializedModel;
            using (var file = File.OpenRead(modelPath))
            {
                serializedModel = ml.Model.Load(file, out var serializedSchema);
                TestCommon.CheckSameSchemas(data.Schema, serializedSchema);
            }
However, the `Save() `method will hit an exception because `Save.IsColumnSavable(type)` returns false. The root cause is we have a customized dataviewtype, which is neither `VectorDataViewType `nor `PrimitiveDataViewType`, the check failed so no columns is added for save. Do you have any guidance on this?

Reply:
> > Part of the `DTRootCauseLocalizationTransformer` class is methods that save the model to a file, and that load the model back from a file. This test should exercise these methods, by using the `ml.Model.Save` and `ml.Model.Load` APIs, and checking that the values returned by the deserialized model are the same as the ones returned by the original model.
> > In reply to: [399045379](https://github.com/dotnet/machinelearning/pull/4925#discussion_r399045379) [](ancestors = 399045379)
> 
> I have tried to use this logic to save and load model,
> 
> ```
>         var modelPath = "temp.zip";
>         //Save model to a file
>         ml.Model.Save(model, data.Schema, modelPath);
> 
>         //Load model from a file
>         ITransformer serializedModel;
>         using (var file = File.OpenRead(modelPath))
>         {
>             serializedModel = ml.Model.Load(file, out var serializedSchema);
>             TestCommon.CheckSameSchemas(data.Schema, serializedSchema);
>         }
> ```
> 
> However, the `Save() `method will hit an exception because `Save.IsColumnSavable(type)` returns false. The root cause is we have a customized dataviewtype, which is neither `VectorDataViewType `nor `PrimitiveDataViewType`, the check failed so no columns is added for save. Do you have any guidance on this?

Has updated according to your suggestion in email discussion